stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -1558,6 +1558,14 @@ function det(A::AbstractMatrix{T}) where T
 end
 det(x::Number) = x
 
+# Resolve Issue #40128
+function det(A::AbstractMatrix{BigInt})
+    m, n = size(A)
+    if m == n || throw(DimensionMismatch("matrix is not square: dimensions are $(size(A))"))
```

Comment:
Seems like this test should be in `det_bareiss`.

Note also that you can simply call `checksquare(A)`.

Reply:
It shouldn't matter whether someone calls `det_bareiss!` directly, or through `det_bareiss`, or through `det` — in all cases, it should check whether the matrix is square.